schnorr_fun/src/frost/chilldkg.rs

```diff
@@ -0,0 +1,1256 @@
+    pub struct Coordinator {
+        threshold: u32,
+        inputs: BTreeMap<u32, Option<KeygenInput>>,
+    }
+
+    impl Coordinator {
+        /// Creates a new coordinator with:
+        ///
+        /// - `threshold`: of key we're trying to generate
+        /// - `n_contributor`: The number of [`Contributor`]s
+        pub fn new(threshold: u32, n_contributor: u32) -> Self {
+            assert!(threshold > 0);
+            Self {
+                threshold,
+                inputs: (0..n_contributor).map(|i| (i, None)).collect(),
+            }
+        }
+
+        /// Adds an `input` from a [`Contributor`].
+        ///
+        /// Note verifying this is the correct input from the correct party is up to your application!
+        pub fn add_input<H: Hash32, NG>(
+            &mut self,
+            schnorr: &Schnorr<H, NG>,
+            from: u32,
+            input: KeygenInput,
+        ) -> Result<(), &'static str> {
+            let entry = match self.inputs.get_mut(&from) {
+                Some(maybe_input) => match maybe_input {
+                    Some(_) => return Err("we already have input from this party"),
+                    none => none,
+                },
+                None => return Err("no input expected from this party"),
+            };
+            if input.com.len() != self.threshold as usize {
+                return Err("input has the wrong threshold");
+            }
+
+            let (first_coeff_even_y, _) = input.com[0].into_point_with_even_y();
+            if !schnorr.verify(&first_coeff_even_y, Message::<Public>::empty(), &input.pop) {
+                return Err("☠ pop didn't verify");
+            }
+            *entry = Some(input);
+
+            Ok(())
+        }
+
+        /// Which [`Contributor`]s are we missing input from.
+        pub fn missing_from(&self) -> BTreeSet<u32> {
+            self.inputs
+                .iter()
+                .filter_map(|(index, input)| match input {
+                    None => Some(*index),
+                    Some(_) => None,
+                })
+                .collect()
+        }
+
+        /// Has the coordinator received input from each [`Contributor`].
+        pub fn is_finished(&self) -> bool {
+            self.inputs.values().all(|v| v.is_some())
+        }
+
+        /// Try and finish input aggregation step.
+        ///
+        /// Returns `None` if [`is_finished`] returns `false`.
+        ///
+        /// [`is_finished`]: Self::is_finished
+        pub fn finish(self) -> Option<AggKeygenInput> {
+            if !self.is_finished() {
+                return None;
+            }
+            let inputs = self.inputs.into_values().flatten().collect::<Vec<_>>();
+            let key_contrib = inputs
+                .iter()
+                .map(|message| (message.com[0], message.pop))
+                .collect();
+            let mut sum_coms = vec![Point::<NonNormal, Public, _>::zero(); self.threshold as usize];
```

Comment:
Could be worth mentioning the trick here, that the coordinator can aggregate all the coefficients except for the first coefficients, since they need to be shared so that the proof-of-possessions can be independently verified.

Reply:
I made an explanation of why the first coefficients aren't aggregated.